### PR TITLE
test: add test with a flashloan

### PR DESCRIPTION
**Pull request by barrutko**

### CODE REVIEW

**Review by hexonaut — DISMISSED** [review]
LGTM

File: test/Integration.t.sol [hunk 1]
```diff
@@ -477,4 +477,106 @@ contract ConcreteTests is CapAutomatorIntegrationTestsBase {
         vm.stopPrank();
     }
 
+    function test_E2E_flashloan() public {
+        DataTypes.ReserveData memory initialReserveData = pool.getReserveData(WBTC);
+
+        // Confirm initial state
+        assertEq(ERC20(WBTC).decimals(), 8);
+        assertEq(currentATokenSupply(initialReserveData), 750);
+        assertEq(initialReserveData.configuration.getSupplyCap(), 3_000);
```
Comment (lucas-manuel) [inline test/Integration.t.sol]: Alignment for all assertEq blocks

File: test/Integration.t.sol [hunk 2]
```diff
@@ -477,4 +477,106 @@ contract ConcreteTests is CapAutomatorIntegrationTestsBase {
...
+        uint256 amount,
+        uint256,
+        address,
+        bytes calldata
+    ) external returns (bool) {
+        // Supply additional funds to the pool, bring the supply closer to the cap than the gap
+        ERC20(asset).approve(address(pool), 40e8);
+        pool.supply(WBTC, 40e8, address(this), 0);
+
+        // Confirm prievious cap before the update
```
Comment (lucas-manuel) [inline test/Integration.t.sol]: previous

File: test/Integration.t.sol [hunk 3]
```diff
@@ -477,4 +477,106 @@ contract ConcreteTests is CapAutomatorIntegrationTestsBase {
...
+        // initialSupply + gap = 750 + 50 = 800
+        assertEq(pool.getReserveData(WBTC).configuration.getSupplyCap(), 800);
+
+        // Confirm the cap can be correctly increased
+        // initialSupply + supppliedFunds + gap = 750 + 40 + 50 = 840
```
Comment (lucas-manuel) [inline test/Integration.t.sol]: supplied

File: test/Integration.t.sol [hunk 4]
```diff
@@ -477,4 +477,106 @@ contract ConcreteTests is CapAutomatorIntegrationTestsBase {
...
+        capAutomator.execSupply(WBTC);
+        assertEq(pool.getReserveData(WBTC).configuration.getSupplyCap(), 840);
+
+        // Supply more funds to attempt a second cap increase
+        ERC20(asset).approve(address(pool), 40e8);
+        pool.supply(WBTC, 40e8, address(this), 0);
+
+        // Confirm the cap cannot be increased twice
+        // initialSupply + totalSupppliedFunds + gap = 750 + 80 + 50 = 880
```
Comment (lucas-manuel) [inline test/Integration.t.sol]: supplied

File: test/Integration.t.sol [hunk 5]
```diff
@@ -477,4 +477,106 @@ contract ConcreteTests is CapAutomatorIntegrationTestsBase {
...
+        capAutomator.execSupply(WBTC);
+        assertEq(pool.getReserveData(WBTC).configuration.getSupplyCap(), 840);
+
+        // Withdraw funds to pay back the flashloan
+        pool.withdraw(WBTC, 80e8, address(this));
+
+        // Confirm the cap cannot be decreased in the same block, even thoough the supply came back to the initial state
+        // initialSupply + totalSupppliedFunds - withdrawnFunds + gap = 750 + 80 - 80 + 50 = 800
```
Comment (lucas-manuel) [inline test/Integration.t.sol]: supplied

File: test/Integration.t.sol [hunk 6]
```diff
@@ -477,4 +477,106 @@ contract ConcreteTests is CapAutomatorIntegrationTestsBase {
...
+
+        // Confirm the cap cannot be increased twice
+        // initialSupply + totalSupppliedFunds + gap = 750 + 80 + 50 = 880
+        capAutomator.execSupply(WBTC);
+        assertEq(pool.getReserveData(WBTC).configuration.getSupplyCap(), 840);
+
+        // Withdraw funds to pay back the flashloan
+        pool.withdraw(WBTC, 80e8, address(this));
+
+        // Confirm the cap cannot be decreased in the same block, even thoough the supply came back to the initial state
```
Comment (lucas-manuel) [inline test/Integration.t.sol]: though

File: test/Integration.t.sol [hunk 7]
```diff
@@ -477,4 +477,106 @@ contract ConcreteTests is CapAutomatorIntegrationTestsBase {
...
+        ERC20(WBTC).approve(address(pool), 40e8);
+        pool.supply(WBTC, 40e8, address(this), 0);
+
+        // Confirm that after some time, smaller than cooldwon, cap cannot be increased
+        capAutomator.execSupply(WBTC);
+        assertEq(pool.getReserveData(WBTC).configuration.getSupplyCap(), 800);
+
+        skip(6 hours + 1);
+
+        // Confirm that after suffcient time passes, cap can be increased again
```
Comment (lucas-manuel) [inline test/Integration.t.sol]: sufficient

File: test/Integration.t.sol [hunk 8]
```diff
@@ -477,4 +477,106 @@ contract ConcreteTests is CapAutomatorIntegrationTestsBase {
...
+        assertEq(pool.getReserveData(WBTC).configuration.getSupplyCap(), 800);
+
+        vm.roll(block.number + 1);
+        skip(6 hours);
+
+        deal(WBTC, address(this), 40e8);
+        ERC20(WBTC).approve(address(pool), 40e8);
+        pool.supply(WBTC, 40e8, address(this), 0);
+
+        // Confirm that after some time, smaller than cooldwon, cap cannot be increased
```
Comment (lucas-manuel) [inline test/Integration.t.sol]: cooldown

File: test/Integration.t.sol [hunk 9]
```diff
@@ -477,4 +477,106 @@ contract ConcreteTests is CapAutomatorIntegrationTestsBase {
...
+    function executeOperation(
+        address asset,
+        uint256 amount,
+        uint256,
+        address,
+        bytes calldata
+    ) external returns (bool) {
+        // Supply additional funds to the pool, bring the supply closer to the cap than the gap
+        ERC20(asset).approve(address(pool), 40e8);
+        pool.supply(WBTC, 40e8, address(this), 0);
```
Comment (lucas-manuel) [inline test/Integration.t.sol]: For this example I think we should supply the max amount to demo the max increase possible with this function

Comment (lucas-manuel) [reply test/Integration.t.sol]: My understanding is its cap + gap but we should show that

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Looking good, some comments

**Review by hexonaut — CHANGES_REQUESTED** [review]
New changes look good. One unrelated, but small thing is I want to set the `ICapAutomator` pragma to `pragma solidity >=0.8.0;`. I think we can just include it in this PR for simplicity.

Interfaces should always be way more flexible.

File: test/Integration.t.sol (hunk 9 above)
Comment (barrutko) [reply test/Integration.t.sol]: In order to do that, I'd need to `deal` some more tokens to `address(this)` to have enough to supply to full and also the numbers would be much uglier and less round. I think this test tests the logic it should test, while edge cases are tested elsewhere. But if you have a strong opinion here, I can add these changes.

File: test/Integration.t.sol (hunk 2 above)
Comment (lucas-manuel) [inline test/Integration.t.sol]: previous

**Review by lucas-manuel — APPROVED** [review]

**Review by hexonaut — APPROVED** [review]
LGTM
